madara-test-runner/src/node.rs

```diff
@@ -0,0 +1,119 @@
+use std::env;
+use std::fmt::Debug;
+use std::fs::{create_dir_all, File};
+use std::net::TcpListener;
+use std::path::{Path, PathBuf};
+use std::process::{Child, Command, ExitStatus, Stdio};
+use std::str::FromStr;
+
+use url::Url;
+
+use crate::Settlement;
+
+const MIN_PORT: u16 = 49_152;
+const MAX_PORT: u16 = 65_535;
+
+#[derive(Debug)]
+/// A wrapper over the Madara process handle, reqwest client and request counter
+///
+/// When this struct goes out of scope, it's `Drop` impl
+/// will take care of killing the Madara process.
+pub struct MadaraNode {
+    process: Child,
+    port: u16,
+}
+
+impl Drop for MadaraNode {
+    fn drop(&mut self) {
+        if let Err(e) = self.process.kill() {
+            eprintln!("Could not kill Madara process: {}", e)
+        }
+    }
+}
+
+fn get_free_port() -> u16 {
+    for port in MIN_PORT..=MAX_PORT {
+        if let Ok(listener) = TcpListener::bind(("127.0.0.1", port)) {
+            return listener.local_addr().expect("No local addr").port();
+        }
+        // otherwise port is occupied
+    }
+    panic!("No free ports available");
+}
+
+fn get_repository_root() -> PathBuf {
+    let manifest_path = Path::new(&env!("CARGO_MANIFEST_DIR"));
+    let repository_root = manifest_path.parent().expect("Failed to get parent directory of CARGO_MANIFEST_DIR");
+    repository_root.to_path_buf()
+}
+
+impl MadaraNode {
+    fn cargo_run(root_dir: &Path, options: Vec<Option<String>>) -> Child {
```

Comment:
rustdoc + should it always be release by default?

Reply:
let's keep it that way for now

Reply:
just precise it in the comment then cause it's not obvious

Reply:
done